### DISCUSSION TITLE
Add a way to specify ligatures for Bitmap fonts

### DISCUSSION THREAD

**Discussion by JustCozzie** [godotengine/godot-proposals#8001] [GUI]
**Describe the project you are working on:**
GBC style pixel-perfect RPG

**Describe the problem or limitation you are having in your project:**
![image](https://github.com/godotengine/godot-proposals/assets/17277922/ce258310-3762-4266-9de5-cddc7bedd62d)
![image](https://github.com/godotengine/godot-proposals/assets/17277922/5ea206ef-6dbc-48bc-bd4e-1e1735a4ebcd)

characters that would traditionally be combined in fixed width fonts (such as 's or 'd) instead take up two full character widths and it is not possible to use different open/close speech marks. At a low render resolution space for text is very limited (typically half of the above example) so with this limitation it becomes more difficult to fit text and often results in uneven spacing.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Ligatures for Bitmap fonts would allow the user to specify sequences that should be combined into a chosen character. This allows for space to be saved, helping prevent text from overflowing available space. It could also be used to create special rules such as:

- Space + " = open speech mark
- ! + Space = condensed exclamation mark (to preserve consistent spacing)

that would currently only be possible by pre-processing all text to replace instances of those combinations occurring with an unused ASCII code - creating friction between writing and implementation.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
![image](https://github.com/godotengine/godot-proposals/assets/17277922/d77ab884-caf8-48c1-93ef-ae0810be86b1)

The import editor currently specifies character type based on a string array, with each character being indexed from the top left and assigned to an ASCII code. This process already supports ranges using a hyphen ('a'-'z'), in my proposal it would also support using a + ('X'+'P') to define that this one character is assigned to that _combination_ and order of ASCII codes.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
While I think it would see good use by the portion of people developing pixel/retro style games, it would also not take any effort to work around. Those who find it unnecessary would simply not specify any ligatures in the font importer.

**Is there a reason why this should be core and not an add-on in the asset library?:**
The Bitmap font system already requires outside resources to use efficiently (ASCII mapper/BMFont) so I think adding to that would be a bad idea. Probably more practically it seems too small in scope to get an addon, as the ones available favour large reworks/overhauls.

**Comment by bruvzg** [comment]
True ligatures are handled by the external shaping library (HarfBuzz) and adding support for it to the bitmap font will likely require creating at least a full OpenType GSUB tables for such fonts.

But, while it's not the same and not exposed to the image import settings (can be easily added), you can set kerining pairs for any type of font from the script (and should be sufficient for any spacing adjustments).
```gdscript
	var f = load("res://325.jpg")
	f.set_kerning(0, f.get_fixed_size(), Vector2i(0x74, 0x21), Vector2(-4, 0)) # t + !
	f.set_kerning(0, f.get_fixed_size(), Vector2i(0x21, 0x31), Vector2(-4, 0)) # ! + 1
```
<img width="287" alt="Screenshot 2023-10-06 at 08 03 51" src="https://github.com/godotengine/godot-proposals/assets/7645683/6a95041d-15d1-4916-bae6-c2370e977b6b">

Adding a more advanced image importer which allows setting arbitrary texture rects and advances for each glyph should be possible as well.

**Reply by Mickeon** [reply]
I cannot fetch the issue I opened last time right now, but this discussion is tangentially related to it.

Edit: https://github.com/godotengine/godot/issues/82254
